src/genworld.cpp

```diff
@@ -337,3 +341,167 @@ void GenerateWorld(GenWorldMode mode, uint size_x, uint size_y, bool reset_setti
+			feature.at("name").get_to(town.name);
+			SetDParamStr(0, town.name);
+			ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_TOWN_FORMATTED_INCORRECTLY, WL_ERROR);
+			return;
+		}
+
+		/* Set town properties. */
+		feature.at("name").get_to(town.name);
+		feature.at("population").get_to(town.population);
+		feature.at("city").get_to(town.is_city);
+
+		/* Set town coordinates. */
+		feature.at("x").get_to(town.x_proportion);
+		feature.at("y").get_to(town.y_proportion);
+
+		/* Check for improper coordinates and warn the player. */
+		if (town.x_proportion <= 0.0f || town.y_proportion <= 0.0f || town.x_proportion >= 1.0f || town.y_proportion >= 1.0f) {
+			SetDParamStr(0, town.name);
+			ShowErrorMessage(STR_TOWN_DATA_ERROR_LOAD_FAILED, STR_TOWN_DATA_ERROR_BAD_COORDINATE, WL_ERROR);
+			return;
+		}
+
+		/* Find the target tile for the town. */
+		TileIndex tile = TileXY(town.x_proportion * Map::MaxX(), town.y_proportion * Map::MaxY());
+
+		/* Try founding on the target tile, and if that doesn't work, find the nearest suitable tile up to 16 tiles away.
+		 * The target might be on water, blocked somehow, or on a steep slope that can't be terraformed by the founding command. */
+		TileIndex search_tile = tile;
+		bool success = CircularTileSearch(&search_tile, 16, 0, 0, TryFoundTownNearby, &town);
+
+		/* If we still fail to found the town, we'll create a sign at the intended location and tell the player how many towns we failed to create in an error message.
+		 * This allows the player to diagnose a heightmap misalignment, if towns end up in the sea, or place towns manually, if in rough terrain. */
+		if (!success) {
+			Command<CMD_PLACE_SIGN>::Post(tile, town.name);
+			failed_towns++;
+			continue;
+		}
+
+		towns.emplace_back(std::make_pair(Town::Get(town.town_id), town.population));
+	}
+
+	/* If we couldn't found a town (or multiple), display a message to the player with the number of failed towns. */
+	if (failed_towns > 0) {
+		SetDParam(0, failed_towns);
+		ShowErrorMessage(STR_TOWN_DATA_ERROR_FAILED_TO_FOUND_TOWN, INVALID_STRING_ID, WL_WARNING);
+	}
+
+	/* Now that we've created the towns, let's grow them to their target populations. */
+	for (const auto &item : towns) {
+		Town *t = item.first;
+		uint population = item.second;
+
+		/* Grid towns can grow almost forever, but the town growth algorithm gets less and less efficient as it wanders roads randomly,
+		 * so we set an arbitrary limit. With a flat map and a 3x3 grid layout this results in about 4900 houses, or 2800 houses with "Better roads." */
+		int try_limit = 1000;
+
+		/* If a town repeatedly fails to grow, continuing to try only wastes time. */
+		int fail_limit = 10;
+
+		/* Grow by a constant number of houses each time, instead of growth based on current town size.
+		 * We want our try limit to apply in a predictable way, no matter the road layout and other geography. */
+		const int HOUSES_TO_GROW = 10;
```

Comment:
One final nitpick: should we move these consts up like other files, and make `1000` and `10` also consts (next to these variables, ofc), so we get `HOUSE_GROW_TRY_LIMIT` and `HOUSE_GROW_FAIL_LIMIT`?

Btw, excellent comment around `try_limit`, exactly addresses my concern/issue with it :)

Reply:
I tried that, and it was messy. These only apply to this function and aren't shared by anything else in the file, so I personally like them as specific as possible. 

Since you phrased your remark as a question, my answer is "In my opinion, no." If your remark was actually a request, I will happily move them for you. 😄

Reply:
Nope. It was an actual question :) Maybe another dev has another opinion, but as long as you had a reason for it, I am pretty much okay with this :)